doc/examples/type_hints.rst

```diff
@@ -0,0 +1,149 @@
+
+.. _type_hints-example:
+
+Type Hints
+===========
+
+As of version 4.1, PyMongo ships with `type hints`_.
+
+With type hints, Python type checkers can easily find bugs before they reveal themselves in your code.  If your IDE is configured to use type hints,
+it can suggest more appropriate completions and highlight errors in your code.
+You can also use the `mypy`_ tool from your command line or in Continuous Integration tests.
+
+All of the public APIs in PyMongo are fully type hinted, and
+several of them support generic parameters for the
+type of document object returned when decoding BSON documents.
+
+Due to `limitations in mypy`_, the default
+values for generic document types are not yet provided (they will eventually be ``Dict[str, any]``).
+
+For a larger set of examples that use types, see the PyMongo `test_mypy module`_.
+
+If you would like to opt out of using the provided types, add the following to
+your `mypy config`_: ::
+
+    [mypy-pymongo]
+    follow_imports = False
+
+
+Basic Usage
+-----------
+
+Note that a type for :class:`~pymongo.mongo_client.MongoClient` must be specified.  Here we use the
+default, unspecified document type:
+
+.. doctest::
+
+  >>> from pymongo import MongoClient
+  >>> client: MongoClient = MongoClient()
+  >>> collection = client.test.test
+  >>> inserted = collection.insert_one({"x": 1, "tags": ["dog", "cat"]})
+  >>> retrieved = collection.find_one({"x": 1})
+  >>> assert isinstance(retrieved, dict)
+
+For a more accurate typing for document type you can use:
+
+.. doctest::
+
+  >>> from typing import Any, Dict
+  >>> from pymongo import MongoClient
+  >>> client: MongoClient[Dict[str, Any]] = MongoClient()
+  >>> collection = client.test.test
+  >>> inserted = collection.insert_one({"x": 1, "tags": ["dog", "cat"]})
+  >>> retrieved = collection.find_one({"x": 1})
+  >>> assert isinstance(retrieved, dict)
+
+Client Document Type
+--------------------
+
+:class:`~pymongo.mongo_client.MongoClient` is generic on the document type used to decode BSON documents.
```

Comment:
We should mention that Client/Database/Collection are all generic too. For instance the TypedDict Movie example would probably make more sense at the collection level:
```python
  >>> from pymongo.collection import Collection
  ...
  >>> client: MongoClient[Dict[str, Any]] = MongoClient()
  >>> collection: Collection[Movie] = client.test.movies
```

Reply:
Done